pil-analyzer/src/type_inference.rs

```diff
@@ -470,10 +472,15 @@ impl<'a> TypeChecker<'a> {
                             self.type_into_substituted(ty)
                         ))
                     })
-            })
-        } else {
-            self.expect_type(&expected_type.ty, expr)
-        }
+            } else {
+                let expected_type = if ty == Type::Tuple(TupleType { items: vec![] }) {
+                    Type::Tuple(TupleType { items: vec![] })
+                } else {
+                    expected_type.ty.clone()
+                };
+                self.expect_type(&expected_type, expr)
```

Comment:
I think we shouldn't call expect_type after calling infer_type on the same expression.

Essentially I would say this should be changed to

```
        self.infer_type_of_expression(expr).and_then(|ty| {
            let ty = self.type_into_substituted(ty);
            let expected_type = if expected_type.allow_array && matches!(ty, Type::Array(_)) {
              ...
            } else if expected_type.allow_empty && ty == Type::Tuple(...)) {
              ...
            } else {
              expected_type.ty.clone()
            };
            self.unifier.unify_types(...)
        })
```